account-kit/smart-contracts/src/ma-v2/account/semiModularAccountV2.ts

```diff
@@ -0,0 +1,149 @@
+import { singleSignerMessageSigner } from "../modules/single-signer-validation/signer.js";
+import { InvalidEntityIdError, InvalidNonceKeyError } from "@aa-sdk/core";
+
+export const DEFAULT_OWNER_ENTITY_ID = 0;
+
+export type SMAV2Account<
+  TSigner extends SmartAccountSigner = SmartAccountSigner
+> = SmartContractAccountWithSigner<"SMAV2Account", TSigner, "0.7.0">;
+
+export type CreateSMAV2AccountParams<
+  TTransport extends Transport = Transport,
+  TSigner extends SmartAccountSigner = SmartAccountSigner
+> = Pick<
+  ToSmartContractAccountParams<"SMAV2Account", TTransport, Chain, "0.7.0">,
+  "transport" | "chain" | "accountAddress"
+> & {
+  signer: TSigner;
+  salt?: bigint;
+  factoryAddress?: Address;
+  initCode?: Hex;
+  initialOwner?: Address;
+  entryPoint?: EntryPointDef<"0.7.0", Chain>;
+} & (
+    | {
+        isGlobalValidation: boolean;
+        entityId: number;
+      }
+    | {
+        isGlobalValidation?: never;
+        entityId?: never;
+      }
+  );
+
+export async function createSMAV2Account<
+  TTransport extends Transport = Transport,
+  TSigner extends SmartAccountSigner = SmartAccountSigner
+>(
+  config: CreateSMAV2AccountParams<TTransport, TSigner>
+): Promise<SMAV2Account<TSigner>>;
+
+export async function createSMAV2Account(
+  config: CreateSMAV2AccountParams
+): Promise<SMAV2Account> {
+  const {
+    transport,
+    chain,
+    signer,
+    salt = 0n,
+    factoryAddress = getDefaultMAV2FactoryAddress(chain),
+    initCode,
+    initialOwner,
+    accountAddress,
+    entryPoint = getEntryPoint(chain, { version: "0.7.0" }),
+    isGlobalValidation = true,
+    entityId = 0,
+  } = config;
+
+  if (entityId >= Number(maxUint32)) {
+    throw new InvalidEntityIdError(entityId);
+  }
+
+  const client = createBundlerClient({
+    transport,
+    chain,
+  });
+
+  const getAccountInitCode = async () => {
+    if (initCode) {
+      return initCode;
+    }
+
+    // If an initial owner is not provided, use the signer's address
+    const ownerAddress = initialOwner ?? (await signer.getAddress());
+
+    return concatHex([
+      factoryAddress,
+      encodeFunctionData({
+        abi: accountFactoryAbi,
+        functionName: "createSemiModularAccount",
+        args: [ownerAddress, salt],
+      }),
+    ]);
+  };
+
+  const baseAccount = await toSmartContractAccount({
+    transport,
+    chain,
+    entryPoint,
+    accountAddress,
+    source: `SMAV2Account`,
+    getAccountInitCode,
+    ...standardExecutor,
+    ...singleSignerMessageSigner(signer),
+  });
+
+  // TODO: add deferred action flag
+  const getAccountNonce = async (nonceKey?: bigint): Promise<bigint> => {
+    const nonceKeySuffix: Hex = `${toHex(entityId, { size: 4 })}${
+      isGlobalValidation ? "01" : "00"
+    }`;
+
+    if (nonceKey && toHex(nonceKey, { size: 5 }) !== nonceKeySuffix) {
+      throw new InvalidNonceKeyError(nonceKey, hexToBigInt(nonceKeySuffix));
+    }
```

Comment:
This is fine, but something that may be a little simpler would be is doing everything with bigints instead of converting everything to hex strings, comparing them, then converting back, which I found a bit confusing especially with the sizes. It might look like
```ts
const nonceKeySuffix = (entityId << 8n)
  + (isGlobalValidation ? 1n : 0n);
if (nonceKey !== nonceKeySuffix) {
  // ...
```
and leaving out all the `hexToBigInt` below.

Actually, I'm confused here: the variable is called `nonceKeySuffix`, but you're requiring `nonceKey` to be equal to `nonceKeySuffix`, not be suffixed by it. If you were intending to compare suffixes via those `size` parameters, that won't work: the `size` param doesn't truncate, it throws if the input is too long. Is this logic correct?

Reply:
> The size param doesn't truncate; it throws if the input is too long
TIL, thought it was like a `slice` haha

During review earlier, aligned with @moldy530 @adam-alchemy to move this to a `uint152 parallelNonce` 
input, appending the `uint40` suffix and using that as the `uint192` nonceKey to the EP. This abstracts the suffix complexity away from devs and they can use smaller numbers such as `0`, `1` for the nonce keys